execute/exectypes/costly_messages.go

```diff
@@ -185,3 +202,63 @@ func (n *StaticMessageExecCostUSD18Calculator) MessageExecCostUSD18(
 }
 
 var _ MessageExecCostUSD18Calculator = &StaticMessageExecCostUSD18Calculator{}
+
+// CCIPMessageFeeUSD18Calculator calculates the fees (paid at source) of a set of messages in USD18s.
+type CCIPMessageFeeUSD18Calculator struct {
+	lggr logger.Logger
+
+	ccipReader readerpkg.CCIPReader
+
+	// RelativeBoostPerWaitHour indicates how much to increase (artificially) the fee paid on the source chain per hour
+	// of wait time, such that eventually the fee paid is greater than the execution cost, and we’ll execute it.
+	// For example: if set to 0.5, that means the fee paid is increased by 50% every hour the message has been waiting.
+	relativeBoostPerWaitHour float64
+
+	now func() time.Time
+}
+
+var _ MessageFeeE18USDCalculator = &CCIPMessageFeeUSD18Calculator{}
+
+// MessageFeeUSD18 Returns a map from message ID to the message's fee in USD18s.
+func (c *CCIPMessageFeeUSD18Calculator) MessageFeeUSD18(
+	ctx context.Context,
+	messages []cciptypes.Message,
+	messageTimeStamps map[cciptypes.Bytes32]time.Time,
+) (map[cciptypes.Bytes32]plugintypes.USD18, error) {
+	linkPriceUSD, err := c.ccipReader.LinkPriceUSD(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get LINK price in USD: %w", err)
+	}
+
+	messageFees := make(map[cciptypes.Bytes32]plugintypes.USD18)
+	for _, msg := range messages {
+		feeUSD18 := big.NewInt(0).Mul(linkPriceUSD.Int, msg.FeeValueJuels.Int)
+		timestamp, ok := messageTimeStamps[msg.Header.MessageID]
+		if !ok {
+			// If a timestamp is missing we can't do fee boosting, but we still record the fee. In the worst case, the
+			// message will not be executed (as it will be considered too costly).
+			c.lggr.Warnw("missing timestamp for message", "messageID", msg.Header.MessageID)
+		} else {
+			feeUSD18 = waitBoostedFee(c.now().Sub(timestamp), feeUSD18, c.relativeBoostPerWaitHour)
+		}
+
+		messageFees[msg.Header.MessageID] = feeUSD18
+	}
+
+	return messageFees, nil
+}
+
+// waitBoostedFee boosts the given fee according to the time passed since the msg was sent.
+// RelativeBoostPerWaitHour is used to normalize the time diff,
+// it makes our loss taking "smooth" and gives us time to react without a hard deadline.
+// At the same time, messages that are slightly underpaid will start going through after waiting for a little bit.
+//
+// wait_boosted_fee(m) = (1 + (now - m.send_time).hours * RELATIVE_BOOST_PER_WAIT_HOUR) * fee(m)
+func waitBoostedFee(waitTime time.Duration, fee *big.Int, relativeBoostPerWaitHour float64) *big.Int {
+	k := 1.0 + waitTime.Hours()*relativeBoostPerWaitHour
+
+	boostedFee := big.NewFloat(0).Mul(big.NewFloat(k), new(big.Float).SetInt(fee))
```

Comment:
Same suggestion as above
```suggestion
	boostedFee := new(big.Float).Mul(big.NewFloat(k), new(big.Float).SetInt(fee))
```

Reply:
Curiously, these two are not equivalent :\